nemo/collections/llm/gpt/model/megatron/hyena/hyena_utils.py

```diff
@@ -1132,6 +1144,136 @@ def sharded_state_dict(self, prefix='', sharded_offsets=(), metadata=None):
+
+        # Extract bias if needed
+        if self._use_conv_bias:
+            bias = self._mixer_module.conv_bias
+            # The mixer bias should have shape [x.shape[1] // 3] for the CUDA kernel
+            if bias.shape[0] != x.shape[1] // 3:
+                # Expand the bias to match the input width
+                bias = bias.repeat_interleave(x.shape[1] // (3 * bias.shape[0]), dim=0)
+        else:
+            # Create a zero bias with the correct shape [x.shape[1] // 3]
+            bias = torch.zeros(x.shape[1] // 3, dtype=x.dtype, device=x.device)
+
+        # Reshape proj_weight if needed (from [groups, channels, kernel_size] to [groups*channels, kernel_size])
+        if proj_weight.dim() == 3:
+            proj_weight = proj_weight.reshape(-1, proj_weight.size(-1))
+        # Reshape mixer_weight if needed (from [groups, channels, kernel_size] to [groups*channels, kernel_size])
+        if mixer_weight.dim() == 3:
+            # If the middle dimension is 1, we can just squeeze it
+            if mixer_weight.size(1) == 1:
+                mixer_weight = mixer_weight.squeeze(1)
+            else:
+                # Otherwise reshape to flatten the first two dimensions
+                mixer_weight = mixer_weight.reshape(-1, mixer_weight.size(-1))
+
+        # maybe handle num_groups
+        proj_weight = proj_weight.repeat_interleave(self._proj_conv_module.group_dim, dim=0)
+        mixer_weight = mixer_weight.repeat_interleave(self._mixer_module.group_dim, dim=0)
+
+        # Support context parallelism similar to how it's done in ParallelCausalDepthwiseConv1d
+        if _use_cp and get_context_parallel_world_size() > 1:
+
+            cp_group = get_context_parallel_group()
+            cp_rank = get_context_parallel_rank()
+
+            # Transfer patches across ranks
+            seq_dim = 2  # Last dimension (L)
+
+            # Get overlapping patches - using the combined effective padding size
+            chunk_a, chunk_b = zigzag_get_overlapping_patches(x, seq_dim=seq_dim, overlap_size=self.effective_pad_size)
+
+            # We're exchanging larger patches once instead of smaller patches twice
+            received_a, received_b = ExchangeOverlappingRegionsCausal.apply(chunk_a, chunk_b, cp_group, cp_rank)
+
+            # Pad and rearrange
+            x = rearrange(x, "b h (nc s) -> (nc b) h s", nc=2)
+            padding = torch.concat([received_a, received_b], dim=0)
+
+            x = torch.concat([padding, x], dim=-1)  # [ncB, D, L]
+            result = self.b2b_causal_conv1d_fn(x, proj_weight, mixer_weight, bias)
+            result = result[..., self.effective_pad_size :]  # Remove padding from output
+            result = rearrange(result, "(nc b) h s -> b h (nc s)", nc=2)
+        else:
+            # Add proper causal padding for the non-CP case
+            x = torch.nn.functional.pad(x, (self.effective_pad_size, 0))
```

Comment:
@farhadrgh I don't think we need padding wo context parallel for b2b convolution.  Am I missing something?

Reply:
We also have a test for equivalency of CP=2 and CP=1 so pretty sure the padding is needed